roachtest: rationalize cluster destruction

andreimatei opened the pull request:
Cluster destruction can be initiated from two places: tests finishing
and a signal being received. Some synchronization is required for
handling races on this destruction. This patch makes this
synchronization nicer. Before this patch, the case where a signal was
received and then a test finishes was handled OK, but the reverse I
think was not.

This patch also kills a global containing a list of all the clusters and
encapsulates it in a nicer clusterRegistry.

Release note: None

andreimatei committed:
roachtest: rationalize cluster destruction

Cluster destruction can be initiated from two places: tests finishing
and a signal being received. Some synchronization is required for
handling races on this destruction. This patch makes this
synchronization nicer. Before this patch, the case where a signal was
received and then a test finishes was handled OK, but the reverse I
think was not.

This patch also kills a global containing a list of all the clusters and
encapsulates it in a nicer clusterRegistry.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35585)
<!-- Reviewable:end -->

andreimatei committed:
roachtest: enforce cluster name uniqueness

To be used in the future where I want to get rid of subtests by
flattening everything and thus producing tests with longer names, at
which point I'll somehow truncate the names of their clusters to the
length limit allowed by GCE and so it'll be more likely to end with
duplicates if nobody's watching.

Release note: None

andreimatei reviewed (commented):
friendly ping

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-:-La77hW0AVGDHVHu5a-g:b-50swta)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)



<!-- Sent from Reviewable.io -->

andreimatei commented:
Bob do you want to be my pal here?

bobvawter reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35585)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @petermattis)

___
*[pkg/cmd/roachtest/cluster.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld65W4M0aLrWYcUhDU7:-Ld65W4M0aLrWYcUhDU8:b-80kozu) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L255)):*
> ```Go
> 	select {
> 	case <-done:
> 	case <-time.After(5 * time.Minute):
> ```

Should this be done by calling `context.WithTimeout` on the `Context` passed into `destroyAllClusters`?
___
*[pkg/cmd/roachtest/cluster.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld66TO0BvzaVHC7EOdC:-Ld66TO0BvzaVHC7EOdD:bjahlum) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L680)):*
> ```Go
> 	// destroyed is used to coordinate between different goroutines that want to
> 	// destroy a cluster.
> 	destroyed chan struct{}
> ```

Maybe change the type here to `<-chan struct{}` since the only useful thing to do here is to `select` or `close()`?
___
*[pkg/cmd/roachtest/cluster.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld665YXB7bSoQxsE6gQ:-Ld665YXB7bSoQxsE6gR:b-udb3bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L695)):*
> ```Go
> 		return false
> 	}
> 	c.destroyState.mu.destroying = true
> ```

It seems odd that a `should` function would mutate the state of the `cluster`; the name sounds idempotent.  Maybe rename it to `beginDestroy` or something and document that this function is a one-shot.
___
*[pkg/cmd/roachtest/cluster.go, line 1072 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld66zhh77eIVbe7JFh1:-Ld66zhh77eIVbe7JFh2:bnvhmyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L1072)):*
> ```Go
> 
> // destroyInner implements the logic for Destroy but without closing c.l.
> func (c *cluster) destroyInner(ctx context.Context) {
> ```

The name of this method is confusing w.r.t. `doDestroy`.  Can the main `Destroy` method defer to a single `doDestroy()` that has a parameter that controls whether or not to close the associated logger?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35585)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter and @petermattis)

___
*[pkg/cmd/roachtest/cluster.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld65W4M0aLrWYcUhDU7:-Ld6A4js-EXmRzrbnEVH:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L255)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Should this be done by calling `context.WithTimeout` on the `Context` passed into `destroyAllClusters`?
</blockquote></details>

done
___
*[pkg/cmd/roachtest/cluster.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld66TO0BvzaVHC7EOdC:-Ld6BPpvEj-uUtdz7SPo:b8xtcna) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L680)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Maybe change the type here to `<-chan struct{}` since the only useful thing to do here is to `select` or `close()`?
</blockquote></details>

well you can't close a receive-only channel. Morally this reference (which is the only reference to this chan) is used both for receiving and for sending (i.e. closing). 
___
*[pkg/cmd/roachtest/cluster.go, line 695 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld665YXB7bSoQxsE6gQ:-Ld6VTjr2Hf-8x0XGUwi:b-2x4j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L695)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

It seems odd that a `should` function would mutate the state of the `cluster`; the name sounds idempotent.  Maybe rename it to `beginDestroy` or something and document that this function is a one-shot.
</blockquote></details>

see now
___
*[pkg/cmd/roachtest/cluster.go, line 1072 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld66zhh77eIVbe7JFh1:-Ld6VPZH73bNV_38OzL-:b-8xx2ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/da895acccdca8171ffd8e0f68c218a7aaf0c0904/pkg/cmd/roachtest/cluster.go#L1072)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

The name of this method is confusing w.r.t. `doDestroy`.  Can the main `Destroy` method defer to a single `doDestroy()` that has a parameter that controls whether or not to close the associated logger?
</blockquote></details>

I don't know what I was drinking when I wrote this. See now.


<!-- Sent from Reviewable.io -->

bobvawter reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-:-Ld8pc9P2pslfIYi7kj9:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @petermattis)

___
*[pkg/cmd/roachtest/cluster.go, line 1166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld8oljyBs6roJvHi01F:-Ld8oljyBs6roJvHi01G:b-7k9srm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7875fbe152b409cb02a1bca286096c22c3203781/pkg/cmd/roachtest/cluster.go#L1166)):*
> ```Go
> 	c.destroyState.mu.Lock()
> 	close(c.destroyState.mu.destroyed)
> 	ch := c.destroyState.mu.destroyed
> ```

Move assignment of ch one line up and call `close(ch)`?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-:-LdAHvyJ1hVTbOnKKJen:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bobvawter and @petermattis)

___
*[pkg/cmd/roachtest/cluster.go, line 1166 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35585#-Ld8oljyBs6roJvHi01F:-LdAHuN95pX4rfPHE-Ol:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7875fbe152b409cb02a1bca286096c22c3203781/pkg/cmd/roachtest/cluster.go#L1166)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Move assignment of ch one line up and call `close(ch)`?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->
